### PR TITLE
GG-21244: Add support for encrypted passwords in Web Agent configuration.

### PR DESCRIPTION
**Pull request by smoldachev**

### CODE REVIEW

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java [hunk 1]
```diff
@@ -324,4 +340,44 @@ public static void send(Session ses, WebSocketResponse evt, long timeout, TimeUn
...
+        return encryptor;
+    }
+
+    /**
+     * @param val Value.
+     */
+    public static String decodeValue(String val) {
+        if (isEncryptedValue(val)) {
+            if (encryptor == null)
+                throw new IgniteException("Failed to decode value, please check that WEB_AGENT_MASTER_PASSWORD env variable is set");
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]:
```suggestion
                throw new IgniteException("Failed to decode value, please check that " + WEB_AGENT_MASTER_PASSWORD_ENV_NAME + "environment variable is set");
```

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java [hunk 2]
```diff
@@ -324,4 +340,44 @@ public static void send(Session ses, WebSocketResponse evt, long timeout, TimeUn
...
+            try {
+                return decrypt(val, encryptor);
+            }
+            catch (EncryptionOperationNotPossibleException e) {
+                throw new IgniteException("Failed to decode value, please check that WEB_AGENT_MASTER_PASSWORD or WEB_AGENT_ENCRYPT_ALGORITHM env variables is correct");
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]:
```suggestion
                throw new IgniteException("Failed to decode value, please check that " + WEB_AGENT_MASTER_PASSWORD_ENV_NAME + " and/or " + WEB_AGENT_ENCRYPT_ALGORITHM_ENV_NAME + " environment variables are correctly specified");
```

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java (hunk 1 above)
Comment (smoldachev) [reply modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]: Fix.

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java (hunk 2 above)
Comment (smoldachev) [reply modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]: Fix.

**Comment by smoldachev** [comment]
> Need to add description to README in agent folder

Please review readme.

File: modules/web-console/web-agent/README.txt [hunk 3]
```diff
@@ -38,6 +38,42 @@ Configuration file:
...
+      Windows: set WEB_AGENT_MASTER_PASSWORD=my-master-password
+
+    2. Select password encryption algorithms.  This step is optional.
+      Supported algorithms:
+        PBEWITHMD5ANDDES - default algorithm
+        PBEWITHMD5ANDTRIPLEDES
+        PBEWITHSHA1ANDDESEDE
+        PBEWITHSHA1ANDRC2_40
+
+    3. Set environment variable with algorithm name.
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/README.txt]:
```suggestion
      Set environment variable with algorithm name.
```

File: modules/web-console/web-agent/README.txt [hunk 4]
```diff
@@ -38,6 +38,42 @@ Configuration file:
...
+      Linux/Mac OS: export  WEB_AGENT_ENCRYPT_ALGORITHM=PBEWITHMD5ANDTRIPLEDES
+      Windows: set  WEB_AGENT_ENCRYPT_ALGORITHM=PBEWITHMD5ANDTRIPLEDES
+
+    4. Encrypt passwords that will be used with Web Agent:
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/README.txt]:
```suggestion
    3. Encrypt passwords that will be used with Web Agent:
```

File: modules/web-console/web-agent/README.txt [hunk 5]
```diff
@@ -38,6 +38,42 @@ Configuration file:
...
+      Linux/Mac OS:  ./encrypt.sh input="plain-text-password" password=my-master-password
+      Windows:  encrypt.bat input="plain-text-password" password=my-master-password
+      The result will be encoded password like this: gKgWcumBHSXZ6C5oVKYVD57u75rNbT1ggF3eUgubZre2zRbqLfV9i8dMdVHKLofh
+
+  And use encrypted passwords with Web Agent in script options or "default.properties" file.
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/README.txt]:
```suggestion
  4. Use encrypted passwords with Web Agent in script options or "default.properties" file.
```

File: modules/web-console/web-agent/README.txt (hunk 4 above)
Comment (smoldachev) [reply modules/web-console/web-agent/README.txt]: Fix.

File: modules/web-console/web-agent/README.txt (hunk 5 above)
Comment (smoldachev) [reply modules/web-console/web-agent/README.txt]: Fix.

File: modules/web-console/web-agent/README.txt [hunk 6]
```diff
@@ -31,13 +31,31 @@ Configuration file:
...
     node-uri=http://10.0.0.1:8080,http://10.0.0.2:8080
 
+Passwords encryption:
+  There several passwords in Web Agent, for example for the SSL connectors. By default, passwords passed as simple text.
+  For security reasons encrypted passwords can be used.
+
+  How to encrypt passwords:
+    1. Run "passwords-key-store-generator.{bat|sh}" and follow instructions.
+
+    2. Add path to passwords key store to default.properties or pass as "ignite-web-console-agent.{bat|sh}" argument.
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/README.txt]:
```suggestion
    2. Add path to key store with passwords to "default.properties" or pass as "ignite-web-console-agent.{bat|sh}" argument.
```

File: modules/web-console/web-agent/bin/passwords-key-store-generator.sh [hunk 7]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/bin/passwords-key-store-generator.sh]: We need bat file also!

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentConfiguration.java [hunk 8]
```diff
@@ -309,7 +319,7 @@ public AgentConfiguration nodeKeyStore(String nodeKeyStore) {
      * @return Node key store password.
      */
     public String nodeKeyStorePassword() {
-        return nodeKeyStorePass;
+        return !F.isEmpty(nodeKeyStorePass) ? nodeKeyStorePass : getPasswordFromKeyStore("node-key-store-password");
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentConfiguration.java]: Lets invert condition to not use negate operator.

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java [hunk 9]
```diff
@@ -324,4 +332,36 @@ public static void send(Session ses, WebSocketResponse evt, long timeout, TimeUn
...
+     * @param name Name.
+     * @param keyStorePath Key store path.
+     * @param keyStorePwd Key store password.
+     */
+    public static String getPasswordFromKeyStore(String name, String keyStorePath, String keyStorePwd) {
+        if (F.isEmpty(keyStorePath))
+            throw new IgniteException("The passwords key store path can't be empty!");
+
+        try {
+            KeyStore ks = KeyStore.getInstance("PKCS12");
```
Comment (nva) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]: Why not JKS?

Comment (smoldachev) [reply modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]: We can't save secret keys in JKS, we can use JCEKS but keytool generate such warning:

The JCEKS keystore uses a proprietary format. It is recommended to migrate to PKCS12 which is an industry standard format using "keytool -importkeystore -srckeystore passwords.p12 -destkeystore passwords.p12 -deststoretype pkcs12".

File: modules/web-console/web-agent/README.txt (hunk 6 above)
Comment (smoldachev) [reply modules/web-console/web-agent/README.txt]: Fix.

File: modules/web-console/web-agent/bin/passwords-key-store-generator.sh (hunk 7 above)
Comment (smoldachev) [reply modules/web-console/web-agent/bin/passwords-key-store-generator.sh]: Wait from @imurchenko

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentConfiguration.java (hunk 8 above)
Comment (smoldachev) [reply modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentConfiguration.java]: Done.

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java [hunk 10]
```diff
@@ -324,4 +332,36 @@ public static void send(Session ses, WebSocketResponse evt, long timeout, TimeUn
...
+                    throw new IgniteException(String.format("Failed to find password in key store: [name=%s, keyStorePath=%s]", name, keyStorePath));
+
+                return new String(secretKey.getEncoded());
+            }
+        }
+        catch (IOException e) {
+            if (e.getCause() instanceof UnrecoverableKeyException)
+                throw new IgniteException("Failed to read password from key store, please check key store password!", e);
+
+            throw new IgniteException("Fail to open passwords key store by path: " + keyStorePath, e);
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]:
```suggestion
            throw new IgniteException("Failed to open passwords key store: " + keyStorePath, e);
```

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java [hunk 11]
```diff
@@ -324,4 +332,36 @@ public static void send(Session ses, WebSocketResponse evt, long timeout, TimeUn
...
     }
+
+    /**
+     * @param name Name.
+     * @param keyStorePath Key store path.
+     * @param keyStorePwd Key store password.
+     */
+    public static String getPasswordFromKeyStore(String name, String keyStorePath, String keyStorePwd) {
+        if (F.isEmpty(keyStorePath))
+            throw new IgniteException("The passwords key store path can't be empty!");
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]:
```suggestion
            throw new IgniteException("Empty path to key store with passwords");
```

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java [hunk 12]
```diff
@@ -324,4 +332,36 @@ public static void send(Session ses, WebSocketResponse evt, long timeout, TimeUn
...
+            }
+        }
+        catch (IOException e) {
+            if (e.getCause() instanceof UnrecoverableKeyException)
+                throw new IgniteException("Failed to read password from key store, please check key store password!", e);
+
+            throw new IgniteException("Fail to open passwords key store by path: " + keyStorePath, e);
+        }
+        catch (GeneralSecurityException e) {
+            throw new IgniteException("Failed to read password from key store!", e);
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]:
```suggestion
            throw new IgniteException("Failed to read password from key store", e);
```

File: modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java [hunk 13]
```diff
@@ -324,4 +332,36 @@ public static void send(Session ses, WebSocketResponse evt, long timeout, TimeUn
...
+
+                if (secretKey == null)
+                    throw new IgniteException(String.format("Failed to find password in key store: [name=%s, keyStorePath=%s]", name, keyStorePath));
+
+                return new String(secretKey.getEncoded());
+            }
+        }
+        catch (IOException e) {
+            if (e.getCause() instanceof UnrecoverableKeyException)
+                throw new IgniteException("Failed to read password from key store, please check key store password!", e);
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/main/java/org/apache/ignite/console/agent/AgentUtils.java]:
```suggestion
                throw new IgniteException("Failed to read password from key store, please check key store password", e);
```

File: modules/web-console/web-agent/src/test/java/org/apache/ignite/testsuites/IgniteWebAgentTestSuite.java [hunk 14]
```diff
@@ -27,7 +29,9 @@
 @RunWith(Suite.class)
 @Suite.SuiteClasses({
     RestExecutorSelfTest.class,
-    SerializationTests.class
+    SerializationTests.class,
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/src/test/java/org/apache/ignite/testsuites/IgniteWebAgentTestSuite.java]: Please sort tests alphabetically.

File: modules/web-console/web-agent/bin/passwords-key-store-generator.sh [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# Define functions
+createKeystore() {
+    keytool -importpassword -alias ${1} -keystore passwords.p12 -storetype pkcs12 -storepass ${2} -keypass ${2}
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-agent/bin/passwords-key-store-generator.sh]: Please add detection of JAVA_HOME in the same way as in ignite.sh, for example.

File: modules/web-console/web-agent/src/test/java/org/apache/ignite/testsuites/IgniteWebAgentTestSuite.java (hunk 14 above)
Comment (smoldachev) [reply modules/web-console/web-agent/src/test/java/org/apache/ignite/testsuites/IgniteWebAgentTestSuite.java]: Done.

**Pull request merged**